File: sdk/typescript/test/builder.test.ts
import { describe, expect, it } from "vitest";
import {
  NETWORK_HOOK_ACTION_MUTATE,
  NETWORK_HOOK_PHASE_AFTER,
  Sandbox,
  VFS_HOOK_ACTION_BLOCK,
  VFS_HOOK_OP_CREATE,
  VFS_HOOK_PHASE_BEFORE,
} from "../src";

describe("Sandbox builder", () => {
  it("sets image on construction", () => {
    const opts = new Sandbox("alpine:latest").options();
    expect(opts.image).toBe("alpine:latest");
  });

  it("chains resource options", () => {
    const opts = new Sandbox("img")
      .withCPUs(4)
      .withMemory(1024)
      .withDiskSize(2048)
      .withTimeout(300)
      .options();

    expect(opts.cpus).toBe(4);
    expect(opts.memoryMb).toBe(1024);
    expect(opts.diskSizeMb).toBe(2048);
    expect(opts.timeoutSeconds).toBe(300);
  });

  it("builds kernel config", () => {
    const opts = new Sandbox("img").withKernel("file:///tmp/vmlinux").options();

    expect(opts.kernelRef).toBe("file:///tmp/vmlinux");
  });

  it("builds network options", () => {
    const opts = new Sandbox("img")
      .allowHost("api.openai.com", "pypi.org")
      .addHost("api.internal", "10.0.0.10")
      .addSecret("API_KEY", "secret", "api.openai.com")
      .withDNSServers("1.1.1.1", "8.8.8.8")
      .withHostname("sandbox-host")
      .withNetworkMTU(1200)
      .blockPrivateIPs()
      .options();

    expect(opts.allowedHosts).toEqual(["api.openai.com", "pypi.org"]);
    expect(opts.addHosts).toEqual([{ host: "api.internal", ip: "10.0.0.10" }]);
    expect(opts.secrets).toEqual([
      { name: "API_KEY", value: "secret", hosts: ["api.openai.com"] },
    ]);
    expect(opts.dnsServers).toEqual(["1.1.1.1", "8.8.8.8"]);
    expect(opts.hostname).toBe("sandbox-host");
    expect(opts.networkMtu).toBe(1200);
    expect(opts.blockPrivateIPs).toBe(true);
    expect(opts.blockPrivateIPsSet).toBe(true);
  });

  it("supports private ip override helpers", () => {
    const allow = new Sandbox("img").allowPrivateIPs().options();
    expect(allow.blockPrivateIPs).toBe(false);
    expect(allow.blockPrivateIPsSet).toBe(true);

    const unset = new Sandbox("img")
      .blockPrivateIPs()
      .unsetBlockPrivateIPs()
      .options();
    expect(unset.blockPrivateIPs).toBe(false);
    expect(unset.blockPrivateIPsSet).toBe(false);
  });

  it("supports no-network option", () => {
    const opts = new Sandbox("img").withNoNetwork().options();
    expect(opts.noNetwork).toBe(true);
  });

  it("builds network interception options", () => {
    const opts = new Sandbox("img")
      .withNetworkInterception({
        rules: [
          {
            phase: NETWORK_HOOK_PHASE_AFTER,
            hosts: ["api.example.com"],
            action: NETWORK_HOOK_ACTION_MUTATE,
            bodyReplacements: [{ find: "foo", replace: "bar" }],
          },
        ],
      })
      .options();

    expect(opts.forceInterception).toBe(true);
    expect(opts.networkInterception?.rules?.[0]).toEqual({
      phase: "after",
      hosts: ["api.example.com"],
      action: "mutate",
      bodyReplacements: [{ find: "foo", replace: "bar" }],
    });
  });

  it("builds mount and env config", () => {
    const opts = new Sandbox("img")
      .withEnv("FOO", "bar")
      .withEnvMap({ BAZ: "qux" })
      .mountHostDir("/data", "/host/data")
      .mountHostDirReadonly("/config", "/host/config")
      .mountMemory("/tmp")
      .mountOverlay("/workspace", "/host/workspace")
      .options();

    expect(opts.env).toEqual({ FOO: "bar", BAZ: "qux" });
    expect(opts.mounts).toEqual({
      "/data": { type: "host_fs", hostPath: "/host/data" },
      "/config": { type: "host_fs", hostPath: "/host/config", readonly: true },
      "/tmp": { type: "memory" },
      "/workspace": { type: "overlay", hostPath: "/host/workspace" },
    });
  });

  it("builds vfs interception and image config", () => {
    const opts = new Sandbox("img")
      .withVFSInterception({
        rules: [
          {
            phase: VFS_HOOK_PHASE_BEFORE,
            ops: [VFS_HOOK_OP_CREATE],
            path: "/workspace/blocked.txt",
            action: VFS_HOOK_ACTION_BLOCK,
          },
        ],
      })
      .withUser("nobody")
      .withEntrypoint("python", "main.py")
      .withImageConfig({ workingDir: "/workspace" })
      .options();

    expect(opts.vfsInterception?.rules?.[0].action).toBe("block");
    expect(opts.imageConfig).toEqual({
      user: "nobody",
      entrypoint: ["python", "main.py"],
      workingDir: "/workspace",
    });
  });

  it("supports port forward builder methods", () => {
    const opts = new Sandbox("img")
      .withPortForward(18080, 8080)
      .withPortForward(18443, 8443)
      .withPortForwardAddresses("127.0.0.1", "0.0.0.0")
      .options();

    expect(opts.portForwards).toEqual([
      { localPort: 18080, remotePort: 8080 },
      { localPort: 18443, remotePort: 8443 },
    ]);
    expect(opts.portForwardAddresses).toEqual(["127.0.0.1", "0.0.0.0"]);
  });
});

File: pkg/sdk/builder_test.go
package sdk

import (
	"context"
	"testing"

	"github.com/jingkaihe/matchlock/pkg/api"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestNew(t *testing.T) {
	b := New("alpine:latest")
	opts := b.Options()
	require.Equal(t, "alpine:latest", opts.Image)
}

func TestBuilderResources(t *testing.T) {
	opts := New("alpine:latest").
		WithCPUs(4).
		WithMemory(2048).
		WithDiskSize(10240).
		WithTimeout(600).
		Options()

	require.Equal(t, 4.0, opts.CPUs)
	require.Equal(t, 2048, opts.MemoryMB)
	require.Equal(t, 10240, opts.DiskSizeMB)
	require.Equal(t, 600, opts.TimeoutSeconds)
}

func TestBuilderKernel(t *testing.T) {
	opts := New("alpine:latest").
		WithKernel("file:///tmp/vmlinux").
		Options()

	require.Equal(t, "file:///tmp/vmlinux", opts.KernelRef)
}

func TestBuilderAllowHost(t *testing.T) {
	opts := New("alpine:latest").
		AllowHost("api.openai.com").
		AllowHost("dl-cdn.alpinelinux.org", "*.github.com").
		Options()

	expected := []string{"api.openai.com", "dl-cdn.alpinelinux.org", "*.github.com"}
	require.Equal(t, expected, opts.AllowedHosts)
}

func TestBuilderAddHost(t *testing.T) {
	opts := New("alpine:latest").
		AddHost("api.internal", "10.0.0.10").
		AddHost("db.internal", "10.0.0.11").
		Options()

	require.Equal(t, []api.HostIPMapping{{Host: "api.internal", IP: "10.0.0.10"}, {Host: "db.internal", IP: "10.0.0.11"}}, opts.AddHosts)
}

func TestBuilderAddSecret(t *testing.T) {
	opts := New("alpine:latest").
		AddSecret("API_KEY", "sk-123", "api.openai.com").
		AddSecretWithPlaceholder("TOKEN", "tok-456", "github_pat_sandbox_placeholder", "*.example.com", "api.example.com").
		Options()

	require.Len(t, opts.Secrets, 2)

	s := opts.Secrets[0]
	assert.Equal(t, "API_KEY", s.Name)
	assert.Equal(t, "sk-123", s.Value)
	require.Len(t, s.Hosts, 1)
	assert.Equal(t, "api.openai.com", s.Hosts[0])

	s = opts.Secrets[1]
	assert.Equal(t, "TOKEN", s.Name)
	assert.Equal(t, "tok-456", s.Value)
	assert.Equal(t, "github_pat_sandbox_placeholder", s.Placeholder)
	require.Len(t, s.Hosts, 2)
}

func TestBuilderBlockPrivateIPs(t *testing.T) {
	opts := New("alpine:latest").BlockPrivateIPs().Options()
	require.True(t, opts.BlockPrivateIPs)
	require.True(t, opts.BlockPrivateIPsSet)
}

func TestBuilderAllowPrivateIPs(t *testing.T) {
	opts := New("alpine:latest").AllowPrivateIPs().Options()
	require.False(t, opts.BlockPrivateIPs)
	require.True(t, opts.BlockPrivateIPsSet)
}

func TestBuilderUnsetBlockPrivateIPs(t *testing.T) {
	opts := New("alpine:latest").
		BlockPrivateIPs().
		UnsetBlockPrivateIPs().
		Options()
	require.False(t, opts.BlockPrivateIPs)
	require.False(t, opts.BlockPrivateIPsSet)
}

func TestBuilderWorkspace(t *testing.T) {
	opts := New("alpine:latest").WithWorkspace("/home/user/code").Options()
	require.Equal(t, "/home/user/code", opts.Workspace)
}

func TestBuilderEnv(t *testing.T) {
	opts := New("alpine:latest").
		WithEnv("FOO", "bar").
		WithEnv("HELLO", "world").
		Options()

	require.Equal(t, map[string]string{
		"FOO":   "bar",
		"HELLO": "world",
	}, opts.Env)
}

func TestBuilderEnvMapMerge(t *testing.T) {
	opts := New("alpine:latest").
		WithEnv("FOO", "old").
		WithEnvMap(map[string]string{
			"FOO": "new",
			"BAR": "baz",
		}).
		Options()

	require.Equal(t, "new", opts.Env["FOO"])
	require.Equal(t, "baz", opts.Env["BAR"])
}

func TestBuilderDNSServers(t *testing.T) {
	opts := New("alpine:latest").
		WithDNSServers("1.1.1.1", "1.0.0.1").
		Options()

	require.Equal(t, []string{"1.1.1.1", "1.0.0.1"}, opts.DNSServers)
}

func TestBuilderDNSServersChained(t *testing.T) {
	opts := New("alpine:latest").
		WithDNSServers("1.1.1.1").
		WithDNSServers("8.8.8.8").
		Options()

	require.Equal(t, []string{"1.1.1.1", "8.8.8.8"}, opts.DNSServers)
}

func TestBuilderDNSServersDefaultEmpty(t *testing.T) {
	opts := New("alpine:latest").Options()
	require.Empty(t, opts.DNSServers)
}

func TestBuilderHostname(t *testing.T) {
	opts := New("alpine:latest").
		WithHostname("override.internal").
		Options()

	require.Equal(t, "override.internal", opts.Hostname)
}

func TestBuilderNetworkMTU(t *testing.T) {
	opts := New("alpine:latest").
		WithNetworkMTU(1200).
		Options()

	require.Equal(t, 1200, opts.NetworkMTU)
}

func TestBuilderNoNetwork(t *testing.T) {
	opts := New("alpine:latest").
		WithNoNetwork().
		Options()

	require.True(t, opts.NoNetwork)
}

func TestBuilderNetworkInterception(t *testing.T) {
	opts := New("alpine:latest").
		WithNetworkInterception().
		Options()

	require.True(t, opts.ForceInterception)
}

func TestBuilderNetworkInterceptionRules(t *testing.T) {
	cfg := &NetworkInterceptionConfig{
		Rules: []NetworkHookRule{
			{
				Phase:  NetworkHookPhaseBefore,
				Action: NetworkHookActionBlock,
				Hosts:  []string{"example.com"},
			},
		},
	}
	opts := New("alpine:latest").
		WithNetworkInterception(cfg).
		Options()

	require.True(t, opts.ForceInterception)
	require.NotNil(t, opts.NetworkInterception)
	require.Len(t, opts.NetworkInterception.Rules, 1)
	assert.Equal(t, "example.com", opts.NetworkInterception.Rules[0].Hosts[0])
}

func TestBuilderPortForwards(t *testing.T) {
	opts := New("alpine:latest").
		WithPortForward(18080, 8080).
		WithPortForward(18443, 8443).
		WithPortForwardAddresses("127.0.0.1", "0.0.0.0").
		Options()

	require.Equal(t, []api.PortForward{
		{LocalPort: 18080, RemotePort: 8080},
		{LocalPort: 18443, RemotePort: 8443},
	}, opts.PortForwards)
	require.Equal(t, []string{"127.0.0.1", "0.0.0.0"}, opts.PortForwardAddresses)
}

func TestBuilderMounts(t *testing.T) {
	opts := New("alpine:latest").
		MountHostDir("/data", "/host/data").
		MountHostDirReadonly("/config", "/host/config").
		MountMemory("/tmp/scratch").
		MountOverlay("/workspace", "/host/workspace").
		Options()

	require.Len(t, opts.Mounts, 4)

	m := opts.Mounts["/data"]
	assert.Equal(t, api.MountTypeHostFS, m.Type)
	assert.Equal(t, "/host/data", m.HostPath)
	assert.False(t, m.Readonly)

	m = opts.Mounts["/config"]
	assert.Equal(t, api.MountTypeHostFS, m.Type)
	assert.Equal(t, "/host/config", m.HostPath)
	assert.True(t, m.Readonly)

	m = opts.Mounts["/tmp/scratch"]
	assert.Equal(t, api.MountTypeMemory, m.Type)

	m = opts.Mounts["/workspace"]
	assert.Equal(t, api.MountTypeOverlay, m.Type)
	assert.Equal(t, "/host/workspace", m.HostPath)
}

func TestBuilderFullChain(t *testing.T) {
	opts := New("python:3.12-alpine").
		WithCPUs(2).
		WithMemory(1024).
		WithEnv("PLAIN_TOKEN", "abc123").
		AllowHost("dl-cdn.alpinelinux.org", "api.anthropic.com").
		AddHost("api.internal", "10.0.0.10").
		AddSecret("ANTHROPIC_API_KEY", "sk-ant-xxx", "api.anthropic.com").
		BlockPrivateIPs().
		WithWorkspace("/code").
		MountHostDirReadonly("/data", "/host/data").
		WithTimeout(120).
		Options()

	require.Equal(t, "python:3.12-alpine", opts.Image)
	require.Equal(t, 2.0, opts.CPUs)
	require.Equal(t, 1024, opts.MemoryMB)
	require.Len(t, opts.AllowedHosts, 2)
	require.Len(t, opts.AddHosts, 1)
	require.Len(t, opts.Secrets, 1)
	require.Equal(t, "abc123", opts.Env["PLAIN_TOKEN"])
	require.True(t, opts.BlockPrivateIPs)
	require.True(t, opts.BlockPrivateIPsSet)
	require.Equal(t, "/code", opts.Workspace)
	require.Len(t, opts.Mounts, 1)
	require.Equal(t, 120, opts.TimeoutSeconds)
}

func TestBuilderFractionalCPUs(t *testing.T) {
	opts := New("alpine:latest").WithCPUs(0.5).Options()
	require.Equal(t, 0.5, opts.CPUs)
}

func TestBuilderVFSInterception(t *testing.T) {
	cfg := &VFSInterceptionConfig{
		Rules: []VFSHookRule{
			{
				Phase:  VFSHookPhaseBefore,
				Ops:    []VFSHookOp{VFSHookOpCreate},
				Path:   "/workspace/blocked.txt",
				Action: VFSHookActionBlock,
			},
		},
	}

	opts := New("alpine:latest").WithVFSInterception(cfg).Options()
	require.NotNil(t, opts.VFSInterception)
	require.Len(t, opts.VFSInterception.Rules, 1)
	assert.Equal(t, "block", opts.VFSInterception.Rules[0].Action)
}

func TestBuilderVFSInterceptionCallback(t *testing.T) {
	cfg := &VFSInterceptionConfig{
		Rules: []VFSHookRule{
			{
				Phase: VFSHookPhaseAfter,
				Ops:   []VFSHookOp{VFSHookOpWrite},
				Path:  "/workspace/*",
				Hook: func(ctx context.Context, event VFSHookEvent) error {
					return nil
				},
			},
		},
	}

	opts := New("alpine:latest").WithVFSInterception(cfg).Options()
	require.NotNil(t, opts.VFSInterception)
	require.Len(t, opts.VFSInterception.Rules, 1)
	assert.NotNil(t, opts.VFSInterception.Rules[0].Hook)
}
